### ISSUE TITLE
[MODULES-383] parent of ModuleClassLoader different between Java 8 and 9+

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#600]
@Ladicek opened **[MODULES-383](https://issues.redhat.com/browse/MODULES-383?redirect=false)** and commented

I'm not 100% sure if it's a bug, but it's a behavioral difference between Java 8 and 9+ that actually breaks one of our (Thorntail) use cases.

So: on Java 8, the parent classloader of each `ModuleClassLoader` is the classloader that loaded JBoss Modules themselves. This is actually even specified in the javadoc of `ConcurrentClassLoader` constructors. In Java 9+, the parent classloader is always the system classloader.

My understanding is that JBoss Modules are typically loaded from system classloader, so there should be no difference most of the time, but in one of our (Thorntail) use cases, we load JBoss Modules from a special classloader \[1\]. This works just fine on Java 8, but breaks on Java 9+.

So I'd like to ask: is there a specific reason why this behavior changes between Java 8 and 9+? I know it's been a long time, but I couldn't find anything useful in Git history. I was also able to verify that when reverting to the original behavior for Java 9+ fixes the Thorntail usecase.

\[1\] For in-IDE usage, we have a special `main` method that spawns a new process that builds an uberjar, after that process finishes, we grab the uberjar and create a `URLClassLoader` on top of it. From that special classloader, we invoke our regular `main` method, which then loads JBoss Modules etc. However, JBoss Modules can't find a bunch of resources, because

---
No further details from [MODULES-383](https://issues.redhat.com/browse/MODULES-383?redirect=false)

**Comment by dmlloyd** [comment]
@Ladicek commented

CC [dlloyd@redhat.com](https://issues.redhat.com/secure/ViewProfile.jspa?name=dlloyd@redhat.com)   - not sure who's the correct contact person re JBoss Modules now :-)

**Comment by dmlloyd** [comment]
@dmlloyd commented

Still me :)

Relying on the identity of the parent class loader is generally not going to be reliable.  In order to add a dependency on the special class loader, a specific dependency type should be used.

**Comment by dmlloyd** [comment]
@Ladicek commented

Thanks!

I think I would probably be able to, somehow, add a dependency on the special classloader to all modules that we load (hmmm...), but that's just half of it. I'd also have to get rid of the classpath (i.e., from the system classloader, only retain the platform classloader stuff). That is probably possible using resource filtering or something like that, but at this point, it feels somewhat brittle.

**Comment by dmlloyd** [comment]
@Ladicek commented

-I just realized I forgot to mention one important detail. We're using the classloader of one of our modules to locate `module.xml` files in one of our `ModuleFinder`-s. That's where things fail right now, and I'm pretty sure that adding a dependency on the special classloader won't help. I think adding the special classloader as a fallback `LocalLoader` might work, except there's no public constructor of `ClassLoaderLocalLoader`. I'll try a few options, but I might end up monkey-patching JBoss Modules to use "our classloader" instead of the system classloader as a parent of the module classloaders.-

-Actually, disregard that. It's true we use the classloader of one of our modules to locate `module.xml` files in one of our `ModuleFinder`-s, but it's not the cause of the problem I'm having right now.-

Actually it's even more complex. I did forget to mention a bunch of important details. So, we have our special `Runner.main` method, which sets up a special `ClassLoader` and invokes our `MainInvoker.main` method from that classloader. This `MainInvoker.main` method loads an "application" module and uses its classloader to execute our regular `Swarm.main`. Since the module -doesn't contain the main class- +has the system classloader as a local loader for some of the paths (which I don't know why)+, it's loaded -from a parent, which unfortunately on JDK 9+ is the system classloader- +from there+. So this is equivalent to just running the `Swarm.main` method from classpath, which we know doesn't work.
